haystack_experimental/components/retrievers/auto_merging_retriever.py

```diff
@@ -0,0 +1,151 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import defaultdict
+from typing import Any, Dict, List
+
+from haystack import DeserializationError, Document, component, default_to_dict
+from haystack.core.serialization import default_from_dict, import_class_by_name
+from haystack.document_stores.types import DocumentStore
+
+
+@component
+class AutoMergingRetriever:
+    """
+    A retriever which returns parent documents of the matched leaf nodes documents, based on a threshold setting.
+
+    The AutoMergingRetriever assumes you have a hierarchical tree structure of documents, where the leaf nodes
+    are indexed in a document store. See the HierarchicalDocumentSplitter for more information on how to create
+    such a structure. During retrieval, if the number of matched leaf documents below the same parent is
+    higher than a defined threshold, the retriever will return the parent document instead of the individual leaf
+    documents.
+
+    The rational is, given that a paragraph is split into multiple chunks represented as leaf documents, and if for
+    a given query, multiple chunks are matched, the whole paragraph might be more informative than the individual
+    chunks alone.
+
+    ```python
+
+    from haystack import Document
+    from haystack_experimental.components.splitters import HierarchicalDocumentBuilder
+    from haystack_experimental.components.retrievers.auto_merging_retriever import AutoMergingRetriever
+    from haystack.document_stores.in_memory import InMemoryDocumentStore
+
+    text = "The sun rose early in the morning. It cast a warm glow over the trees. Birds began to sing."
+
+    docs = [Document(content=text)]
+    builder = HierarchicalDocumentBuilder(block_sizes=[10, 3], split_overlap=0, split_by="word")
+    docs = builder.run(docs)
```

Comment:
we can improve this. Something like:

```suggestion
    original_document = Document(content=text)
    builder = HierarchicalDocumentSplitter(block_sizes=[10, 3], split_overlap=0, split_by="word")
    docs = builder.run([original_document])["documents"]
```